module5/lessons/Week3/Authorization.md

```diff
@@ -0,0 +1,212 @@
+
+var builder = WebApplication.CreateBuilder(args);
+
+// Add services to the container.
+builder.Services.AddControllersWithViews();
+
+builder.Services.AddDbContext<MessageBoardContext>(options =>
+    options.UseNpgsql(builder.Configuration.GetConnectionString("IdentityMessageBoardDb")).UseSnakeCaseNamingConvention());
+
+builder.Services.AddDefaultIdentity<ApplicationUser>(options => 
+    options.SignIn.RequireConfirmedAccount = true)
+        .AddRoles<IdentityRole>() // this is new!
+        .AddEntityFrameworkStores<MessageBoardContext>();
+
+var app = builder.Build();
+
+// Configure the HTTP request pipeline.
+if (!app.Environment.IsDevelopment())
+{
+    app.UseExceptionHandler("/Home/Error");
+}
+app.UseStaticFiles();
+
+app.UseRouting();
+
+app.UseAuthentication();
+app.UseAuthorization(); // make sure you have this line
+
+app.MapControllerRoute(
+    name: "default",
+    pattern: "{controller=Home}/{action=Index}/{id?}");
+
+app.MapRazorPages();
+
+app.Run();
+
+```
+
+And, if you can believe it... that's pretty much all the setup!  Now you are ready to start using Roles!
+
+<section class='call-to-action' markdown='1'>
+
+With a partner, add the following decorator to the AllMessages action:
+
+```c#
+        [Authorize(Roles = "Admin")]
+        public IActionResult AllMessages()
+            ...
+```
+
+Then: 
+* Try to access the All Messages link as a **not logged in user**
+* Try to access the All Messages link as a **logged in user**
+
+How are those experiences different?
+
+Be ready to share out: how would you explain how the **Authorize** attribute controls user behavior?
+
+</section>
+
+<section class='instructor-notes' markdown='1'>
+
+Be sure students are thinking about the order of operations - in terms of auth, authentication is more important (gets checked first) than authorization.
+
+</section>
+
+### Add Roles to Database and Existing Users
+
+In order to see Role based authenticaion in action we need to:
+* Create Roles
+* Assign Roles to Users
+
+In the past, we have created forms to add records to our database throug a web application - why might we not want to do that for roles?  At some point, a role will have to be assigned by a developer.  
+
+Think of an application like Amazon.  If you are a merchant on Amazon, you must be registered with the company.  Once registered, you may be able to create your own employee-users, but the initial setup will always be done by Amazon (or, a developer at Amazon).  So, it will be common to manage Roles from the database - using SQL!
+
+#### Add Admin Role to Database
+
+<section class='call-to-action' markdown='1'>
+
+**Practice your SQL!**  
+
+* Open pgAdmin and connect to the IdentityMessageBoard Database
+* Insert a record into `AspNetRoles`
+    * name should be 'Admin'
+* Add the admin role to a user
+    * Hint: keep a close eye on your ERD! Make sure you are adding records to the correct table.  If you are unsure, ask an instructor!
```

Comment:
What do you think about doing a slack waterfall or something when folks come back to verify that everyone has the right inserts, and if not folks can correct?

Reply:
Also love that this is some very naturally flowing SQL practice!

Reply:
Yea, we do want to make sure everyone has it completed correctly - a waterfall is a great idea.